src/test/java/org/cryptomator/fusecloudaccess/CloudAccessFSTest.java

```diff
@@ -44,26 +61,336 @@ public void testInterrupt() throws InterruptedException {
+		@Test
+		public void testGetAttrSuccess() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.completedFuture(CloudItemMetadataProvider.ofPath(PATH)));
+
+			Assertions.assertEquals(0, cloudFs.getattr(PATH.toString(), fileStat));
+		}
+
+		@DisplayName("getAttr() returns ENOENT when resource is not found.")
+		@Test
+		public void testGetAttrNotFound() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new NotFoundException()));
+
+			Assertions.assertEquals(-ErrorCodes.ENOENT(), cloudFs.getattr(PATH.toString(), fileStat));
+		}
+
+		@DisplayName("getAttr() returns EIO on CloudProviderException.")
+		@Test
+		public void testGetAttrCloudError() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new CloudProviderException()));
+
+			Assertions.assertEquals(-ErrorCodes.EIO(), cloudFs.getattr(PATH.toString(), fileStat));
+		}
+
+		@DisplayName("getAttr() returns EIO on any exception.")
+		@Test
+		public void testGetAttrSomeException() {
+			Mockito.when(provider.itemMetadata(PATH))
+					.thenReturn(CompletableFuture.failedFuture(new Exception()));
+
+			Assertions.assertEquals(-ErrorCodes.EIO(), cloudFs.getattr(PATH.toString(), fileStat));
+		}
+	}
+
+	//@Nested
+	static class ReadDirTests {
+
+		private static final long OFFSET = 0L;
+		private static final List<CloudItemMetadata> ITEM_LISTING = List.of(
+				CloudItemMetadataProvider.ofPath(PATH.resolve("asd")),
+				CloudItemMetadataProvider.ofPath(PATH.resolve("qwe")),
+				CloudItemMetadataProvider.ofPath(PATH.resolve("yxc")),
+				CloudItemMetadataProvider.ofPath(PATH.resolve("justAnother"))
+		);
+
+		private CloudProvider provider;
+		private CloudAccessFS cloudFs;
+		private Pointer buf;
+		private FuseFileInfo fi;
+
+		@BeforeEach
+		public void setup() {
+			provider = Mockito.mock(CloudProvider.class);
+			cloudFs = new CloudAccessFS(provider, CloudAccessFSTest.TIMEOUT);
+			buf = Mockito.mock(Pointer.class);
+			fi = Mockito.mock(FuseFileInfo.class);
+		}
+
+		@DisplayName("Successful readdir() returns 0 and copies all elements")
+		@Test
+		public void testSuccess() {
+			FuseFillDirImpl filler = FuseFillDirImpl.getUnlimitedFiller();
+
+			var expectedListing = new ArrayList<String>();
+			expectedListing.add(".");
+			expectedListing.add("..");
+			expectedListing.addAll(ITEM_LISTING.stream().map(CloudItemMetadata::getName).collect(Collectors.toList()));
+
+			Mockito.when(provider.listExhaustively(PATH))
+					.thenReturn(CompletableFuture.completedFuture(new CloudItemList(ITEM_LISTING)));
+
+			Assertions.assertEquals(0, cloudFs.readdir(PATH.toString(), buf, filler, OFFSET, fi));
+			Assertions.assertIterableEquals(expectedListing, filler.getListing());
+		}
+
+		@DisplayName("readdir() returns ENOMEM if output buffer is full")
+		@Test
+		public void testFullNativeBufferReturnsENOMEM() {
+			FuseFillDirImpl filler = FuseFillDirImpl.getENOMEMFiller();
```

Comment:
Can we use a mocked filler instead?

Reply:
resolved by 580e8b613afce1e69aa3085df73644e6849eecc7